scheduled_tasks/tasks.py

```diff
@@ -57,37 +119,145 @@ async def sync_auth0_roles():
     logger.info(f"Found {len(roles)} roles")
 
     db_session = next(get_db_session())
-    with db_session.begin():
+    auth0_role_ids: set[str] = set()
+    try:
         for role in roles:
             logger.info(f"  Role: {role.name}")
+            auth0_role_ids.add(role.id)
             db_role = db_session.get(Auth0Role, role.id)
-            if db_role is not None:
-                logger.info("    Role already exists in DB")
-                continue
+            created = False
+            restored = False
+            if db_role is None:
+                db_role = Auth0Role.get_deleted_by_id(db_session, role.id)
+                if db_role is not None:
+                    restored = True
+                    db_role.restore(db_session, commit=False)
+                else:
+                    created = True
+                    db_role = Auth0Role(
+                        id=role.id,
+                        name=role.name,
+                        description=role.description,
+                    )
+            db_session.add(db_role)
+            if created:
+                logger.info("    Role created in DB")
+            elif restored:
+                logger.info("    Role restored from soft delete")
             else:
-                logger.info("    Role does not exist in DB, creating")
-                db_role = Auth0Role(
-                    id=role.id,
-                    name=role.name,
-                    description=role.description
+                logger.info("    Role exists in DB, updating fields if necessary")
+            if db_role.name != role.name or db_role.description != role.description:
+                db_role.name = role.name
+                db_role.description = role.description
+        # Soft delete roles missing from Auth0
+        db_session.flush()
+        existing_roles = db_session.exec(select(Auth0Role)).all()
+        for db_role in existing_roles:
+            if db_role.id not in auth0_role_ids:
+                logger.info(f"    Soft deleting role {db_role.name} ({db_role.id}) absent from Auth0")
+                db_role.delete(db_session, commit=False)
+        db_session.commit()
+    finally:
+        db_session.close()
+
+
+async def sync_auth0_user_roles():
```

Comment:
This could definitely do with some refactoring/splitting up into different functions. Otherwise we are not going to remember what it does.